lib/SPIRVProducerPass.cpp

```diff
@@ -235,9 +236,37 @@ struct SPIRVInstruction {
 
 struct SPIRVProducerPassImpl {
 
+  // Struct to handle generation of ArrayStride decorations.
+  // It supports two styles of stride generation to avoid unnecessarily
+  // perturbing the instruction order generated during the transition to opaque
+  // pointers (see #816).
+  // TODO(#816): remove the type in this struct and update tests as necessary.
+  struct StrideType {
+    Type *type;
+    uint32_t stride;
+    SPIRVID id;
+
+    StrideType(Type *type, uint32_t stride, SPIRVID id)
+        : type(type), stride(stride), id(id) {}
+    StrideType(Type *type) : type(type), stride(0), id(0) {}
+
+    bool operator<(const StrideType &x) const {
+      if (type < x.type)
```

Comment:
It feels like it has to be. The uint is likely what happens under the hood, but you can put pointers in std::set which is ordered based on less than. Also that link has the following later on:

>If a pointer p compare equal to pointer q, p<=q and p>=q both yield true and p<q and p>q both yield false.
>If a pointer p compares greater than a pointer q, then p>=q, p>q, q<=p, and q<p all yield true and p<=q, p<q, q>=p, and q>p all yield false.

This to me implies its defined. Separately, I had to define operator< for the UniqueVector (which is implemented using std::map<T, unsigned>)`, but previously we were just storing the pointer in it so this would have been pre-existing.